rust/tw_tests/tests/chains/ethereum/ethereum_abi.rs

```diff
@@ -191,3 +191,102 @@ fn test_ethereum_abi_decode_value() {
     assert!(output.error_message.is_empty());
     assert_eq!(output.param_str, "42");
 }
+
+#[test]
+fn test_ethereum_get_function_signature() {
+    let abi = r#"{"constant":false,"inputs":[{"name":"_to","type":"address"},{"name":"_value","type":"uint256"}],"name":"transfer","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"}"#;
+
+    let abi_string = TWStringHelper::create(abi);
+
+    let actual = TWStringHelper::wrap(unsafe {
+        tw_ethereum_abi_get_function_signature(CoinType::Ethereum as u32, abi_string.ptr())
+    })
+    .to_string()
+    .expect("!tw_ethereum_abi_function_get_signature returned nullptr");
+
+    assert_eq!(actual, "transfer(address,uint256)");
+}
+
+#[test]
+fn test_ethereum_get_function_signature_complex() {
+    // From: https://docs.soliditylang.org/en/latest/abi-spec.html#handling-tuple-types
+    let abi = r#"
+{
+  "inputs": [
+    {
+      "components": [
+        {
+          "internalType": "uint256",
+          "name": "a",
+          "type": "uint256"
+        },
+        {
+          "internalType": "uint256[]",
+          "name": "b",
+          "type": "uint256[]"
+        },
+        {
+          "components": [
+            {
+              "internalType": "uint256",
+              "name": "x",
+              "type": "uint256"
+            },
+            {
+              "internalType": "uint256",
+              "name": "y",
+              "type": "uint256"
+            }
+          ],
+          "internalType": "struct Test.T[]",
+          "name": "c",
+          "type": "tuple[]"
+        }
+      ],
+      "internalType": "struct Test.S",
+      "name": "",
+      "type": "tuple"
+    },
+    {
+      "components": [
+        {
+          "internalType": "uint256",
+          "name": "x",
+          "type": "uint256"
+        },
+        {
+          "internalType": "uint256",
+          "name": "y",
+          "type": "uint256"
+        }
+      ],
+      "internalType": "struct Test.T",
+      "name": "",
+      "type": "tuple"
+    },
+    {
+      "internalType": "uint256",
+      "name": "",
+      "type": "uint256"
+    }
+  ],
+  "name": "f",
+  "outputs": [],
+  "stateMutability": "pure",
+  "type": "function"
+}
+"#;
+
+    let abi_string = TWStringHelper::create(abi);
+
+    let actual = TWStringHelper::wrap(unsafe {
+        tw_ethereum_abi_get_function_signature(CoinType::Ethereum as u32, abi_string.ptr())
```

Comment:
There is `tw_ethereum_abi_function_get_signature` function already. However, it takes a function name and `Params`.
I think we need to bring them to a common denominator, in particular:
```diff
-// Return the function type signature, of the form "baz(int32,uint256)".
-message FunctionGetTypeInput {
-    // Function signature. Includes function inputs if they are.
-    // Examples:
-    // - `functionName()`
-    // - `functionName()`
-    // - `functionName(bool)`
-    // - `functionName(uint256,bytes32)`
-    string function_name = 1;
+// Return the function type signature, of the form "transfer(address,uint256)".
+message FunctionGetSignatureInput {
+    message FunctionNameParams {
+        // Function name. For example, "baz".
+        string function_name = 1;
+
+        // A set of ABI type parameters.
+        repeated Param inputs = 2;
+    }
 
-    // A set of ABI type parameters.
-    repeated Param inputs = 2;
+    oneof abi {
+        // A pair of function name and parameters.
+        FunctionNameParams function_name_params = 1;
+
+        // A set of ABI parameters in JSON.
+        // Expected to be a JSON array at the entry level.
+        // Example:
+        // ```
+        // {
+        //     "constant": false,
+        //     "inputs": [
+        //         {
+        //             "name": "_to",
+        //             "type": "address"
+        //         },
+        //         {
+        //             "name": "_value",
+        //             "type": "uint256"
+        //         }
+        //     ],
+        //     "name": "transfer",
+        //     "outputs": [],
+        //     "payable": false,
+        //     "stateMutability": "nonpayable",
+        //     "type": "function"
+        // }
+        // ```
+        string abi_json = 2;
+    }
 }

```

Reply:
Please let me know what do you think about the refactoring? Note that `tw_ethereum_abi_function_get_signature` is not exposed into Kotlin, Swift and other bindings, it's only used internally from C++, so we can freely change it as no backward compatibility will be broken

Reply:
To make the old and new code appear more cohesive and clear, I can make a few small adjustments to the old code:

1. Rename tw_ethereum_abi_function_get_signature to tw_ethereum_abi_function_get_type.
2. Rename EvmEntryExt.get_abi_function_signature to EvmEntryExt.get_function_signature_from_proto.

Please let me know what you think about the adjustments. I would appreciate your feedback.

Reply:
Makes sense. It's ok, let's follow the first way then 👍